In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_excel('Online retail.xlsx' , header= None)
df

0
0     shrimp,almonds,avocado,vegetables mix,green gr...
1                                burgers,meatballs,eggs
2                                               chutney
3                                        turkey,avocado
4     mineral water,milk,energy bar,whole wheat rice...
...                                                 ...
7496                      butter,light mayo,fresh bread
7497  burgers,frozen vegetables,eggs,french fries,ma...
7498                                            chicken
7499                                 escalope,green tea
7500    eggs,frozen smoothie,yogurt cake,low fat yogurt

[7501 rows x 1 columns]

In [3]:
df1 = df.iloc[:,0].str.split(',' , expand = True)
df1

0                  1            2                 3   \
0            shrimp            almonds      avocado    vegetables mix   
1           burgers          meatballs         eggs              None   
2           chutney               None         None              None   
3            turkey            avocado         None              None   
4     mineral water               milk   energy bar  whole wheat rice   
...             ...                ...          ...               ...   
7496         butter         light mayo  fresh bread              None   
7497        burgers  frozen vegetables         eggs      french fries   
7498        chicken               None         None              None   
7499       escalope          green tea         None              None   
7500           eggs    frozen smoothie  yogurt cake    low fat yogurt   

                4                 5     6               7             8   \
0     green grapes  whole weat flour  yams  cottage cheese  energy drink   
1             None              None  None            None          None   
2             None              None  None            None          None   
3             None              None  None            None          None   
4        green tea              None  None            None          None   
...            ...               ...   ...             ...           ...   
7496          None              None  None            None          None   
7497     magazines         green tea  None            None          None   
7498          None              None  None            None          None   
7499          None              None  None            None          None   
7500          None              None  None            None          None   

                9               10         11     12     13             14  \
0     tomato juice  low fat yogurt  green tea  honey  salad  mineral water   
1             None            None       None   None   None           None   
2             None            None       None   None   None           None   
3             None            None       None   None   None           None   
4             None            None       None   None   None           None   
...            ...             ...        ...    ...    ...            ...   
7496          None            None       None   None   None           None   
7497          None            None       None   None   None           None   
7498          None            None       None   None   None           None   
7499          None            None       None   None   None           None   
7500          None            None       None   None   None           None   

          15                 16               17       18         19  
0     salmon  antioxydant juice  frozen smoothie  spinach  olive oil  
1       None               None             None     None       None  
2       None               None             None     None       None  
3       None               None             None     None       None  
4       None               None             None     None       None  
...      ...                ...              ...      ...        ...  
7496    None               None             None     None       None  
7497    None               None             None     None       None  
7498    None               None             None     None       None  
7499    None               None             None     None       None  
7500    None               None             None     None       None  

[7501 rows x 20 columns]

apriori library requires Dataset as a list of list
here, the dataset is a big list and each transaction in the dataset is an inner list witin the outher big list.
so, convert pandas dataframe into a list of lists.



In [3]:
!pip install apyori

In [5]:
from apyori import apriori

In [6]:
df1.shape

(7501, 20)

In [7]:
records = []
for i in range(0,7501):
    records.append([str(df1.values[i,j]) for j in range (0, 20)])

In [8]:
records[0:2]


[['shrimp',
  'almonds',
  'avocado',
  'vegetables mix',
  'green grapes',
  'whole weat flour',
  'yams',
  'cottage cheese',
  'energy drink',
  'tomato juice',
  'low fat yogurt',
  'green tea',
  'honey',
  'salad',
  'mineral water',
  'salmon',
  'antioxydant juice',
  'frozen smoothie',
  'spinach',
  'olive oil'],
 ['burgers',
  'meatballs',
  'eggs',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None',
  'None']]

In [9]:
#apply apriori algorithm
association_rules = apriori(records,min_support = 0.0045 , min_confidence = 0.2 , min_lift= 3)
association_results = list(association_rules)

In [10]:
association_results[1]


RelationRecord(items=frozenset({'escalope', 'mushroom cream sauce'}), support=0.005732568990801226, ordered_statistics=[OrderedStatistic(items_base=frozenset({'mushroom cream sauce'}), items_add=frozenset({'escalope'}), confidence=0.3006993006993007, lift=3.790832696715049)])

items_base= antecedent, 

items_add = consequent items
mushroom cream sauce --> escalope
1. support = 0.0057 implies that approximately 0.573% of all transactions contain both escalope and mushroom cream sauce.
2. confidence = 0.30 implies that there is approximately a 30.07% chance that if a transaction contains mushroom cream sauce, it also contains escalope.
3.  lift - 3.79 implies that transactions containing mushroom cream sauce are approximately 3.79 times more likely to also contain escalope 

In [11]:
association_results = sorted(association_results, key = lambda x: x[1], reverse = True)

In [12]:
for item in association_results:
    pair = item.items
    items = [x for x in pair]
    # Process each OrderedStatistic object
    for stat in item.ordered_statistics:
        antecedent = list(stat.items_base)
        consequent = list(stat.items_add)
        
        rule = str(antecedent) + ' -> ' + str(consequent)
        
        print('Rule: ' + rule)
        print('Support: ' + str(item.support))
        print('Confidence: ' + str(stat.confidence))
        print('Lift: ' + str(stat.lift))
        print('=============================================')

Rule: ['herb & pepper'] -> ['ground beef']
Support: 0.015997866951073192
Confidence: 0.3234501347708895
Lift: 3.2919938411349285
Rule: ['herb & pepper'] -> ['None', 'ground beef']
Support: 0.015997866951073192
Confidence: 0.3234501347708895
Lift: 3.2919938411349285
Rule: ['None', 'herb & pepper'] -> ['ground beef']
Support: 0.015997866951073192
Confidence: 0.3234501347708895
Lift: 3.2919938411349285
Rule: ['frozen vegetables', 'spaghetti'] -> ['ground beef']
Support: 0.008665511265164644
Confidence: 0.31100478468899523
Lift: 3.165328208890303
Rule: ['frozen vegetables', 'spaghetti'] -> ['None', 'ground beef']
Support: 0.008665511265164644
Confidence: 0.31100478468899523
Lift: 3.165328208890303
Rule: ['frozen vegetables', 'None', 'spaghetti'] -> ['ground beef']
Support: 0.008665511265164644
Confidence: 0.31100478468899523
Lift: 3.165328208890303
Rule: ['whole wheat pasta'] -> ['olive oil']
Support: 0.007998933475536596
Confidence: 0.2714932126696833
Lift: 4.122410097642296
Rule: ['whole

# or use mlxtend

In [13]:
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules


In [14]:
te = TransactionEncoder()
te_ary = te.fit(records).transform(records)
df2 = pd.DataFrame(te_ary, columns=te.columns_)
df2

asparagus   None  almonds  antioxydant juice  asparagus  avocado  \
0          False  False     True               True      False     True   
1          False   True    False              False      False    False   
2          False   True    False              False      False    False   
3          False   True    False              False      False     True   
4          False   True    False              False      False    False   
...          ...    ...      ...                ...        ...      ...   
7496       False   True    False              False      False    False   
7497       False   True    False              False      False    False   
7498       False   True    False              False      False    False   
7499       False   True    False              False      False    False   
7500       False   True    False              False      False    False   

      babies food  bacon  barbecue sauce  black tea  ...  turkey  \
0           False  False           False      False  ...   False   
1           False  False           False      False  ...   False   
2           False  False           False      False  ...   False   
3           False  False           False      False  ...    True   
4           False  False           False      False  ...   False   
...           ...    ...             ...        ...  ...     ...   
7496        False  False           False      False  ...   False   
7497        False  False           False      False  ...   False   
7498        False  False           False      False  ...   False   
7499        False  False           False      False  ...   False   
7500        False  False           False      False  ...   False   

      vegetables mix  water spray  white wine  whole weat flour  \
0               True        False       False              True   
1              False        False       False             False   
2              False        False       False             False   
3              False        False       False             False   
4              False        False       False             False   
...              ...          ...         ...               ...   
7496           False        False       False             False   
7497           False        False       False             False   
7498           False        False       False             False   
7499           False        False       False             False   
7500           False        False       False             False   

      whole wheat pasta  whole wheat rice   yams  yogurt cake  zucchini  
0                 False             False   True        False     False  
1                 False             False  False        False     False  
2                 False             False  False        False     False  
3                 False             False  False        False     False  
4                 False              True  False        False     False  
...                 ...               ...    ...          ...       ...  
7496              False             False  False        False     False  
7497              False             False  False        False     False  
7498              False             False  False        False     False  
7499              False             False  False        False     False  
7500              False             False  False         True     False  

[7501 rows x 121 columns]

In [15]:
df2.drop( 'None', axis =1 , inplace = True)
df2

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0          False     True               True      False     True        False   
1          False    False              False      False    False        False   
2          False    False              False      False    False        False   
3          False    False              False      False     True        False   
4          False    False              False      False    False        False   
...          ...      ...                ...        ...      ...          ...   
7496       False    False              False      False    False        False   
7497       False    False              False      False    False        False   
7498       False    False              False      False    False        False   
7499       False    False              False      False    False        False   
7500       False    False              False      False    False        False   

      bacon  barbecue sauce  black tea  blueberries  ...  turkey  \
0     False           False      False        False  ...   False   
1     False           False      False        False  ...   False   
2     False           False      False        False  ...   False   
3     False           False      False        False  ...    True   
4     False           False      False        False  ...   False   
...     ...             ...        ...          ...  ...     ...   
7496  False           False      False        False  ...   False   
7497  False           False      False        False  ...   False   
7498  False           False      False        False  ...   False   
7499  False           False      False        False  ...   False   
7500  False           False      False        False  ...   False   

      vegetables mix  water spray  white wine  whole weat flour  \
0               True        False       False              True   
1              False        False       False             False   
2              False        False       False             False   
3              False        False       False             False   
4              False        False       False             False   
...              ...          ...         ...               ...   
7496           False        False       False             False   
7497           False        False       False             False   
7498           False        False       False             False   
7499           False        False       False             False   
7500           False        False       False             False   

      whole wheat pasta  whole wheat rice   yams  yogurt cake  zucchini  
0                 False             False   True        False     False  
1                 False             False  False        False     False  
2                 False             False  False        False     False  
3                 False             False  False        False     False  
4                 False              True  False        False     False  
...                 ...               ...    ...          ...       ...  
7496              False             False  False        False     False  
7497              False             False  False        False     False  
7498              False             False  False        False     False  
7499              False             False  False        False     False  
7500              False             False  False         True     False  

[7501 rows x 120 columns]

In [16]:
frequent_itemsets = apriori(df2, min_support=0.0045, use_colnames = True)
frequent_itemsets

support                                           itemsets
0    0.020397                                          (almonds)
1    0.008932                                (antioxydant juice)
2    0.004666                                        (asparagus)
3    0.033329                                          (avocado)
4    0.004533                                      (babies food)
..        ...                                                ...
837  0.006266       (mineral water, whole wheat rice, spaghetti)
838  0.005066                   (pancakes, olive oil, spaghetti)
839  0.004533        (eggs, chocolate, mineral water, spaghetti)
840  0.004933        (chocolate, mineral water, milk, spaghetti)
841  0.004533  (milk, mineral water, frozen vegetables, spagh...

[842 rows x 2 columns]

In [17]:
rules = association_rules(frequent_itemsets, min_threshold= 0.4)
rules.sort_values('lift' , ascending = False)

antecedents      consequents  antecedent support  \
38         (cooking oil, ground beef)      (spaghetti)            0.008399   
58     (frozen vegetables, olive oil)           (milk)            0.011332   
83              (shrimp, ground beef)      (spaghetti)            0.011465   
55   (frozen vegetables, ground beef)      (spaghetti)            0.016931   
20             (ground beef, chicken)      (spaghetti)            0.009465   
..                                ...              ...                 ...   
25                  (eggs, chocolate)  (mineral water)            0.033196   
34             (chocolate, spaghetti)  (mineral water)            0.039195   
23               (spaghetti, chicken)  (mineral water)            0.017198   
104               (shrimp, spaghetti)  (mineral water)            0.021197   
7                            (salmon)  (mineral water)            0.042528   

     consequent support   support  confidence      lift  leverage  conviction  \
38             0.174110  0.004799    0.571429  3.281995  0.003337    1.927076   
58             0.129583  0.004799    0.423529  3.268410  0.003331    1.509908   
83             0.174110  0.005999    0.523256  3.005315  0.004003    1.732354   
55             0.174110  0.008666    0.511811  2.939582  0.005718    1.691742   
20             0.174110  0.004799    0.507042  2.912193  0.003151    1.675377   
..                  ...       ...         ...       ...       ...         ...   
25             0.238368  0.013465    0.405622  1.701663  0.005552    1.281394   
34             0.238368  0.015865    0.404762  1.698053  0.006522    1.279541   
23             0.238368  0.006932    0.403101  1.691084  0.002833    1.275981   
104            0.238368  0.008532    0.402516  1.688630  0.003479    1.274731   
7              0.238368  0.017064    0.401254  1.683336  0.006927    1.272045   

     zhangs_metric  
38        0.701197  
58        0.701996  
83        0.674995  
55        0.671179  
20        0.662891  
..             ...  
25        0.426498  
34        0.427860  
23        0.415815  
104       0.416635  
7         0.423972  

[113 rows x 10 columns]

#### Assocaitions found: 
 1. There is an association between cooking oil, ground beef, and spaghetti.
 2. There is an association between olive oil, frozen vegetables, and milk.
 3. There is an association between ground beef, shrimp, and spaghetti.
 4. There is an association between eggs, chocolate, and mineral water.
 5. There is an association between chicken, spaghetti, and mineral water.

1. Lift measures how much more likely the consequent is to occur given the antecedent, relative to its general occurrence in the dataset. It provides insight into the strength and direction of the association.Lift measures how much more likely the consequent is to occur given the antecedent, relative to its general occurrence in the dataset. It provides insight into the strength and direction of the association.
2.  Support measures the frequency or proportion of transactions that include both the antecedent and the consequent of a rule.

3.  Confidence measures the likelihood that the consequent of a rule occurs given that the antecedent occurs. It represents the reliability of the rule.

#### What Are Some Limitations or Challenges of Association Rules Mining?
1. Scalability: Increased data size leads to exponential growth in the number of potential itemsets and rules, making computations costly.

2. Data Sparsity: Many itemsets may have low support in large datasets, resulting in sparse and less meaningful rule generation.

3. Rule Interpretation: The high number of rules can be overwhelming, requiring additional filtering to identify the most useful ones.

4. Parameter Sensitivity: Results depend heavily on parameters like min_support and min_confidence, affecting rule quality.

5. Overfitting: High-confidence rules may fit the training data too closely, reducing generalizability to new data.

6. Lack of Temporal Information: Association rules do not capture the order or timing of transactions.

7. Not Suitable for All Types of Data: Association rule mining is less effective for continuous or non-categorical data without preprocessing.

8. Context Dependency: Rules may be meaningful in one context but not in another, making their applicability context-specific.